<html>
  <head>
    <link rel="preconnect" href="https://fonts.googleapis.com">
    <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
    <link href="https://fonts.googleapis.com/css2?family=Roboto:ital,wght@0,100;0,300;0,400;0,500;0,700;0,900;1,100;1,300;1,400;1,500;1,700;1,900&display=swap" rel="stylesheet">

    <style>
      * {
        font-family: Roboto, FontAwesome;
      }

      p {
        margin: 8px;
      }

      body {
        width: 70%;
        margin: auto;
      }

      h1,
      h2,
      h3,
      h4 {
        margin-top: 0.5em;
        margin-bottom: 0.5em;
      }

      .annotator-doc {
        display: flex;
        margin-top: 60px;
        flex-direction: row;
        border: 1px solid rgb(218,218,218);
      }

      .annotator-line {
        font-size: 18px;
        vertical-align: top;
        width: 40%;
        position: sticky;
        position: -webkit-sticky;
        align-self: flex-start;
        top: 0px;
        word-break: break-word;
      }

      .annotator {
        background-color: #eee;
        padding: 10px 8px;
      }

      .annotator-title {
        font-size: 24px;
        font-weight: 500;
      }

      .attribute-list {
        width: 60%;
        border-right: 1px solid rgb(218,218,218);
      }
      
      .attribute {
        padding: 20px 0 20px;
        display: flex;
        flex-direction: row;
        justify-content: space-between;
        border-bottom: 1px solid rgb(218,218,218);
      }

      .attribute:last-child {
        border-bottom: none;
      }

      .attribute-name {
        font-size: 22px;
        font-weight: 500;
        margin: 8px;
        word-break: break-word;
      }

      .attribute-type, .attribute-internal, .resource-id {
        font-size: 18px;
        margin: 8px;
      }

      .attribute-internal {
        font-style: italic;
      }

      .attribute-description {
        font-size: 18px;
        width: 55%;
        margin-right: 20px;
      }

      .attribute-info {
        font-size: 22px;
        align-self: center;
        width: 35%;
      }

      .no-description {
        font-style: italic;
        color: #9c9c9c;
        padding: 10px 8px;
      }

      .markdown {
        padding: 8px;
      }

      details, .resource, .annotator-description {
        padding: 10px 8px;
      }

      img {
        height: auto; 
        width: auto; 
        max-width: 95%; 
        max-height: 95%;
      }
    </style>
  </head>

  <body>
    <h1>Pipeline Documentation</h1>
    


    
    <div class="annotator-doc">
      <div class="attribute-list">
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">phylop100way</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: phyloP100way</div>
            <p>The scores is a number that reflects the conservation at a position.</p>

<p><img src="https://grr.seqpipe.org/hg38/scores/phyloP100way/statistics/histogram_phyloP100way.png" alt="HISTOGRAM" /></p>

<p>small values: less conserved,
large_values more conserved</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

          </div>
        </div>
        
      </div>

      <div class="annotator-line">
        <div class="annotator">
          <div class="annotator-title">Annotator type: position_score</div>
          
          <div class="resource-id">Resource id: <a href="https://grr.seqpipe.org/hg38/scores/phyloP100way/index.html">
                  hg38/scores/phyloP100way</a></div>                            
          
        </div>
  
        
        <div class="annotator-description">
          <ul>
<li><p>Annotator to use with genomic scores depending on genomic position like
phastCons, phyloP, FitCons2, etc.</p></li>
<li><p><a href="https://www.iossifovlab.com/gpfuserdocs/administration/annotation_tools.html#position-score" target="_blank">More info</a></p></li>
</ul>

        </div>
        

        
        <div class="resource">
          <div><b>Resource</b></div>
          <div>id: <a href="https://grr.seqpipe.org/hg38/scores/phyloP100way/index.html">
                  hg38/scores/phyloP100way</a></div>                            
          <div>type: position_score</div>
          
            <div>Description:</div>
            <div class="markdown">
              <p>phyloP (phylogenetic p-values) conservation score based on the 
multiple alignments of 100 vertebrate genomes (including human).</p>

            </div>
          
        </div>
        
      </div>
  
    </div>
    
    <div class="annotator-doc">
      <div class="attribute-list">
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">phylop30way</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: phyloP30way</div>
            <p>The scores is a number that reflects the conservation at a position.</p>

<p><img src="https://grr.seqpipe.org/hg38/scores/phyloP30way/statistics/histogram_phyloP30way.png" alt="HISTOGRAM" /></p>

<p>small values: less conserved,
large_values more conserved</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

          </div>
        </div>
        
      </div>

      <div class="annotator-line">
        <div class="annotator">
          <div class="annotator-title">Annotator type: position_score</div>
          
          <div class="resource-id">Resource id: <a href="https://grr.seqpipe.org/hg38/scores/phyloP30way/index.html">
                  hg38/scores/phyloP30way</a></div>                            
          
        </div>
  
        
        <div class="annotator-description">
          <ul>
<li><p>Annotator to use with genomic scores depending on genomic position like
phastCons, phyloP, FitCons2, etc.</p></li>
<li><p><a href="https://www.iossifovlab.com/gpfuserdocs/administration/annotation_tools.html#position-score" target="_blank">More info</a></p></li>
</ul>

        </div>
        

        
        <div class="resource">
          <div><b>Resource</b></div>
          <div>id: <a href="https://grr.seqpipe.org/hg38/scores/phyloP30way/index.html">
                  hg38/scores/phyloP30way</a></div>                            
          <div>type: position_score</div>
          
            <div>Description:</div>
            <div class="markdown">
              <p>phyloP (phylogenetic p-values) conservation score based on the 
multiple alignments of 29 genome sequences to the human genome.
The higher the more conservative. Scores range from -20.000 to 1.312.</p>

            </div>
          
        </div>
        
      </div>
  
    </div>
    
    <div class="annotator-doc">
      <div class="attribute-list">
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">phylop20way</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: phyloP20way</div>
            <p>The scores is a number that reflects the conservation at a position.</p>

<p><img src="https://grr.seqpipe.org/hg38/scores/phyloP20way/statistics/histogram_phyloP20way.png" alt="HISTOGRAM" /></p>

<p>small values: less conserved,
large_values more conserved</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

          </div>
        </div>
        
      </div>

      <div class="annotator-line">
        <div class="annotator">
          <div class="annotator-title">Annotator type: position_score</div>
          
          <div class="resource-id">Resource id: <a href="https://grr.seqpipe.org/hg38/scores/phyloP20way/index.html">
                  hg38/scores/phyloP20way</a></div>                            
          
        </div>
  
        
        <div class="annotator-description">
          <ul>
<li><p>Annotator to use with genomic scores depending on genomic position like
phastCons, phyloP, FitCons2, etc.</p></li>
<li><p><a href="https://www.iossifovlab.com/gpfuserdocs/administration/annotation_tools.html#position-score" target="_blank">More info</a></p></li>
</ul>

        </div>
        

        
        <div class="resource">
          <div><b>Resource</b></div>
          <div>id: <a href="https://grr.seqpipe.org/hg38/scores/phyloP20way/index.html">
                  hg38/scores/phyloP20way</a></div>                            
          <div>type: position_score</div>
          
            <div>Description:</div>
            <div class="markdown">
              <p>phyloP (phylogenetic p-values) conservation score based on the multiple
alignments of 19 genome sequences to the human genome. The higher the
more conservative. Scores range from from -14.191 to 1.199.</p>

            </div>
          
        </div>
        
      </div>
  
    </div>
    
    <div class="annotator-doc">
      <div class="attribute-list">
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">phylop7way</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: phyloP7way</div>
            <p>The scores is a number that reflects the conservation at a position.</p>

<p><img src="https://grr.seqpipe.org/hg38/scores/phyloP7way/statistics/histogram_phyloP7way.png" alt="HISTOGRAM" /></p>

<p>small values: less conserved,
large_values more conserved</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

          </div>
        </div>
        
      </div>

      <div class="annotator-line">
        <div class="annotator">
          <div class="annotator-title">Annotator type: position_score</div>
          
          <div class="resource-id">Resource id: <a href="https://grr.seqpipe.org/hg38/scores/phyloP7way/index.html">
                  hg38/scores/phyloP7way</a></div>                            
          
        </div>
  
        
        <div class="annotator-description">
          <ul>
<li><p>Annotator to use with genomic scores depending on genomic position like
phastCons, phyloP, FitCons2, etc.</p></li>
<li><p><a href="https://www.iossifovlab.com/gpfuserdocs/administration/annotation_tools.html#position-score" target="_blank">More info</a></p></li>
</ul>

        </div>
        

        
        <div class="resource">
          <div><b>Resource</b></div>
          <div>id: <a href="https://grr.seqpipe.org/hg38/scores/phyloP7way/index.html">
                  hg38/scores/phyloP7way</a></div>                            
          <div>type: position_score</div>
          
            <div>Description:</div>
            <div class="markdown">
              <p>phyloP (phylogenetic p-values) conservation score based on the multiple
alignments of 6 vertebrate genomes to the human genome. The higher the
more conservative. Scores range from from -5.220 to 1.062.</p>

            </div>
          
        </div>
        
      </div>
  
    </div>
    
    <div class="annotator-doc">
      <div class="attribute-list">
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">phastcons100way</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: phastCons100way</div>
            <p>The scores is a number that reflects the conservation at a position.</p>

<p><img src="https://grr.seqpipe.org/hg38/scores/phastCons100way/statistics/histogram_phastCons100way.png" alt="HISTOGRAM" /></p>

<p>small values: less conserved,
large_values more conserved</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

          </div>
        </div>
        
      </div>

      <div class="annotator-line">
        <div class="annotator">
          <div class="annotator-title">Annotator type: position_score</div>
          
          <div class="resource-id">Resource id: <a href="https://grr.seqpipe.org/hg38/scores/phastCons100way/index.html">
                  hg38/scores/phastCons100way</a></div>                            
          
        </div>
  
        
        <div class="annotator-description">
          <ul>
<li><p>Annotator to use with genomic scores depending on genomic position like
phastCons, phyloP, FitCons2, etc.</p></li>
<li><p><a href="https://www.iossifovlab.com/gpfuserdocs/administration/annotation_tools.html#position-score" target="_blank">More info</a></p></li>
</ul>

        </div>
        

        
        <div class="resource">
          <div><b>Resource</b></div>
          <div>id: <a href="https://grr.seqpipe.org/hg38/scores/phastCons100way/index.html">
                  hg38/scores/phastCons100way</a></div>                            
          <div>type: position_score</div>
          
            <div>Description:</div>
            <div class="markdown">
              <p>phastCons100way is a conservation score based on a multiple alignments 
of 99 vertebrate genomes to the human genome.
The larger the score, the more conserved the site. Scores range from 0 to 1.</p>

            </div>
          
        </div>
        
      </div>
  
    </div>
    
    <div class="annotator-doc">
      <div class="attribute-list">
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">phastcons30way</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: phastCons30way</div>
            <p>The scores is a number that reflects the conservation at a position.</p>

<p><img src="https://grr.seqpipe.org/hg38/scores/phastCons30way/statistics/histogram_phastCons30way.png" alt="HISTOGRAM" /></p>

<p>small values: less conserved,
large_values more conserved</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

          </div>
        </div>
        
      </div>

      <div class="annotator-line">
        <div class="annotator">
          <div class="annotator-title">Annotator type: position_score</div>
          
          <div class="resource-id">Resource id: <a href="https://grr.seqpipe.org/hg38/scores/phastCons30way/index.html">
                  hg38/scores/phastCons30way</a></div>                            
          
        </div>
  
        
        <div class="annotator-description">
          <ul>
<li><p>Annotator to use with genomic scores depending on genomic position like
phastCons, phyloP, FitCons2, etc.</p></li>
<li><p><a href="https://www.iossifovlab.com/gpfuserdocs/administration/annotation_tools.html#position-score" target="_blank">More info</a></p></li>
</ul>

        </div>
        

        
        <div class="resource">
          <div><b>Resource</b></div>
          <div>id: <a href="https://grr.seqpipe.org/hg38/scores/phastCons30way/index.html">
                  hg38/scores/phastCons30way</a></div>                            
          <div>type: position_score</div>
          
            <div>Description:</div>
            <div class="markdown">
              <p>phastCons30way is a conservation score based on a multiple alignments
of 29 genomes to the human genome.
The larger the score, the more conserved the site. Scores range from 0 to 1.</p>

            </div>
          
        </div>
        
      </div>
  
    </div>
    
    <div class="annotator-doc">
      <div class="attribute-list">
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">phastcons20way</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: phastCons20way</div>
            <p>The scores is a number that reflects the conservation at a position.</p>

<p><img src="https://grr.seqpipe.org/hg38/scores/phastCons20way/statistics/histogram_phastCons20way.png" alt="HISTOGRAM" /></p>

<p>small values: less conserved,
large_values more conserved</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

          </div>
        </div>
        
      </div>

      <div class="annotator-line">
        <div class="annotator">
          <div class="annotator-title">Annotator type: position_score</div>
          
          <div class="resource-id">Resource id: <a href="https://grr.seqpipe.org/hg38/scores/phastCons20way/index.html">
                  hg38/scores/phastCons20way</a></div>                            
          
        </div>
  
        
        <div class="annotator-description">
          <ul>
<li><p>Annotator to use with genomic scores depending on genomic position like
phastCons, phyloP, FitCons2, etc.</p></li>
<li><p><a href="https://www.iossifovlab.com/gpfuserdocs/administration/annotation_tools.html#position-score" target="_blank">More info</a></p></li>
</ul>

        </div>
        

        
        <div class="resource">
          <div><b>Resource</b></div>
          <div>id: <a href="https://grr.seqpipe.org/hg38/scores/phastCons20way/index.html">
                  hg38/scores/phastCons20way</a></div>                            
          <div>type: position_score</div>
          
            <div>Description:</div>
            <div class="markdown">
              <p>phastCons20way is a conservation score based on a multiple alignments
of 19 vertebrate genomes to the human genome.
The larger the score, the more conserved the site. Scores range from 0 to 1.</p>

            </div>
          
        </div>
        
      </div>
  
    </div>
    
    <div class="annotator-doc">
      <div class="attribute-list">
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">phastcons7way</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: phastCons7way</div>
            <p>The scores is a number that reflects the conservation at a position.</p>

<p><img src="https://grr.seqpipe.org/hg38/scores/phastCons7way/statistics/histogram_phastCons7way.png" alt="HISTOGRAM" /></p>

<p>small values: less conserved,
large_values more conserved</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

          </div>
        </div>
        
      </div>

      <div class="annotator-line">
        <div class="annotator">
          <div class="annotator-title">Annotator type: position_score</div>
          
          <div class="resource-id">Resource id: <a href="https://grr.seqpipe.org/hg38/scores/phastCons7way/index.html">
                  hg38/scores/phastCons7way</a></div>                            
          
        </div>
  
        
        <div class="annotator-description">
          <ul>
<li><p>Annotator to use with genomic scores depending on genomic position like
phastCons, phyloP, FitCons2, etc.</p></li>
<li><p><a href="https://www.iossifovlab.com/gpfuserdocs/administration/annotation_tools.html#position-score" target="_blank">More info</a></p></li>
</ul>

        </div>
        

        
        <div class="resource">
          <div><b>Resource</b></div>
          <div>id: <a href="https://grr.seqpipe.org/hg38/scores/phastCons7way/index.html">
                  hg38/scores/phastCons7way</a></div>                            
          <div>type: position_score</div>
          
            <div>Description:</div>
            <div class="markdown">
              <p>phastCons7way is a conservation score based on a multiple alignments
of 6 vertebrate genomes to the human genome.
The larger the score, the more conserved the site. Scores range from 0 to 1.</p>

            </div>
          
        </div>
        
      </div>
  
    </div>
    
    <div class="annotator-doc">
      <div class="attribute-list">
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">cadd_raw</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: cadd_raw</div>
            <p>CADD raw score for functional prediction of a SNP. The larger the score 
the more likely the SNP has damaging effect</p>

<p><img src="https://grr.seqpipe.org/hg38/scores/CADD_v1.4/statistics/histogram_cadd_raw.png" alt="HISTOGRAM" /></p>

<p>small values: less damaging,
large_values more damaging</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

<p><strong>nucleotide_aggregator</strong>: <code>max</code> [default]</p>

          </div>
        </div>
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">cadd_phred</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: cadd_phred</div>
            <p>CADD phred-like score. This is phred-like rank score based on whole
genome CADD raw scores. The larger the score the more likely the SNP
has damaging effect.</p>

<p><img src="https://grr.seqpipe.org/hg38/scores/CADD_v1.4/statistics/histogram_cadd_phred.png" alt="HISTOGRAM" /></p>

<p>small values: less damaging,
large_values more damaging</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

<p><strong>nucleotide_aggregator</strong>: <code>max</code> [default]</p>

          </div>
        </div>
        
      </div>

      <div class="annotator-line">
        <div class="annotator">
          <div class="annotator-title">Annotator type: np_score</div>
          
          <div class="resource-id">Resource id: <a href="https://grr.seqpipe.org/hg38/scores/CADD_v1.4/index.html">
                  hg38/scores/CADD_v1.4</a></div>                            
          
        </div>
  
        
        <div class="annotator-description">
          <ul>
<li><p>Annotator to use with genomic scores depending on genomic position and
nucleotide change like CADD, MPC, etc.</p></li>
<li><p><a href="https://www.iossifovlab.com/gpfuserdocs/administration/annotation_tools.html#np-score" target="_blank">More info</a></p></li>
</ul>

        </div>
        

        
        <div class="resource">
          <div><b>Resource</b></div>
          <div>id: <a href="https://grr.seqpipe.org/hg38/scores/CADD_v1.4/index.html">
                  hg38/scores/CADD_v1.4</a></div>                            
          <div>type: np_score</div>
          
            <div>Description:</div>
            <div class="markdown">
              <p>CADD score for functional prediction of a SNP. Please refer to Kircher
et al. (2014) Nature Genetics 46(3):310-5 for details. The larger the score the
more likely the SNP has damaging effect. </p>

            </div>
          
        </div>
        
      </div>
  
    </div>
    
    <div class="annotator-doc">
      <div class="attribute-list">
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">hg19_annotatable</div>
            <div class="attribute-type">Type: annotatable</div>
            
            <div class="attribute-internal">Internal</div>
            
          </div>
          <div class="attribute-description">
            <div>source: liftover_annotatable</div>
            <p>Lifted over allele.</p>

          </div>
        </div>
        
      </div>

      <div class="annotator-line">
        <div class="annotator">
          <div class="annotator-title">Annotator type: liftover_annotator</div>
          
          <div class="resource-id">Resource id: <a href="https://grr.seqpipe.org/liftover/hg38ToHg19/index.html">
                  liftover/hg38ToHg19</a></div>                            
          
          <div class="resource-id">Resource id: <a href="https://grr.seqpipe.org/hg19/genomes/GATK_ResourceBundle_5777_b37_phiX174/index.html">
                  hg19/genomes/GATK_ResourceBundle_5777_b37_phiX174</a></div>                            
          
        </div>
  
        
        <div class="no-description">
          No description
        </div>
        

        
        <div class="resource">
          <div><b>Resource</b></div>
          <div>id: <a href="https://grr.seqpipe.org/liftover/hg38ToHg19/index.html">
                  liftover/hg38ToHg19</a></div>                            
          <div>type: liftover_chain</div>
          
        </div>
        
        <div class="resource">
          <div><b>Resource</b></div>
          <div>id: <a href="https://grr.seqpipe.org/hg19/genomes/GATK_ResourceBundle_5777_b37_phiX174/index.html">
                  hg19/genomes/GATK_ResourceBundle_5777_b37_phiX174</a></div>                            
          <div>type: genome</div>
          
            <div>Description:</div>
            <div class="markdown">
              <h2>HG19 Reference Genome</h2>

<p>Default HG19 reference genome used by GPF</p>

            </div>
          
        </div>
        
      </div>
  
    </div>
    
    <div class="annotator-doc">
      <div class="attribute-list">
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">fitcons_i6_merged</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: fc_i6_score</div>
            <p>probability that a point mutation at each position in a genome will influence fitness</p>

<p><img src="https://grr.seqpipe.org/hg19/scores/FitCons-i6-merged/statistics/histogram_fc_i6_score.png" alt="HISTOGRAM" /></p>

<p>small values: less likely to have fitness consequence,
large_values more likely to have fitness consequence</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

          </div>
        </div>
        
      </div>

      <div class="annotator-line">
        <div class="annotator">
          <div class="annotator-title">Annotator type: position_score</div>
          
          <div class="resource-id">Resource id: <a href="https://grr.seqpipe.org/hg19/scores/FitCons-i6-merged/index.html">
                  hg19/scores/FitCons-i6-merged</a></div>                            
          
        </div>
  
        
        <div class="annotator-description">
          <ul>
<li><p>Annotator to use with genomic scores depending on genomic position like
phastCons, phyloP, FitCons2, etc.</p></li>
<li><p><a href="https://www.iossifovlab.com/gpfuserdocs/administration/annotation_tools.html#position-score" target="_blank">More info</a></p></li>
<li><p><strong>input_annotatable</strong>: <code>hg19_annotatable</code></p></li>
</ul>

        </div>
        

        
        <div class="resource">
          <div><b>Resource</b></div>
          <div>id: <a href="https://grr.seqpipe.org/hg19/scores/FitCons-i6-merged/index.html">
                  hg19/scores/FitCons-i6-merged</a></div>                            
          <div>type: position_score</div>
          
            <div>Description:</div>
            <div class="markdown">
              <p>fitCons score predicts the fraction of genomic positions belonging to a specific
function class (defined by epigenomic "fingerprint") that are under selective
pressure. Scores range from 0 to 1, with a larger score indicating a higher proportion
of nucleic sites of the functional class the genomic position belong to are under
selective pressure, therefore more likely to be functional important. Integrated
(i6) scores are integrated across three cell types (GM12878, H1-hESC and HUVEC).
More details can be found in doi:10.1038/ng.3196.</p>

            </div>
          
        </div>
        
      </div>
  
    </div>
    
    <div class="annotator-doc">
      <div class="attribute-list">
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">linsight</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: Linsight</div>
            <p>The LINSIGHT score measures the probability of negative selection on noncoding
sites.</p>

<p><img src="https://grr.seqpipe.org/hg19/scores/Linsight/statistics/histogram_Linsight.png" alt="HISTOGRAM" /></p>

<p>small values: negative selection is less likely,
large_values negative selection is more likely</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

          </div>
        </div>
        
      </div>

      <div class="annotator-line">
        <div class="annotator">
          <div class="annotator-title">Annotator type: position_score</div>
          
          <div class="resource-id">Resource id: <a href="https://grr.seqpipe.org/hg19/scores/Linsight/index.html">
                  hg19/scores/Linsight</a></div>                            
          
        </div>
  
        
        <div class="annotator-description">
          <ul>
<li><p>Annotator to use with genomic scores depending on genomic position like
phastCons, phyloP, FitCons2, etc.</p></li>
<li><p><a href="https://www.iossifovlab.com/gpfuserdocs/administration/annotation_tools.html#position-score" target="_blank">More info</a></p></li>
<li><p><strong>input_annotatable</strong>: <code>hg19_annotatable</code></p></li>
</ul>

        </div>
        

        
        <div class="resource">
          <div><b>Resource</b></div>
          <div>id: <a href="https://grr.seqpipe.org/hg19/scores/Linsight/index.html">
                  hg19/scores/Linsight</a></div>                            
          <div>type: position_score</div>
          
            <div>Description:</div>
            <div class="markdown">
              <p>LINSIGHT improves the prediction of noncoding nucleotide sites at which
mutations are likely to have deleterious fitness consequences, and which,
therefore, are likely to be phenotypically important. LINSIGHT combines
a generalized linear model for functional genomic data with a probabilistic
model of molecular evolution.</p>

<p>Huang, YF., Gulko, B. &amp; Siepel, A. Fast, scalable prediction of deleterious
noncoding variants from functional and population genomic data.
<em>Nat Genet</em> <strong>49</strong>, 618-624 (2017). https://doi.org/10.1038/ng.3810</p>

            </div>
          
        </div>
        
      </div>
  
    </div>
    
    <div class="annotator-doc">
      <div class="attribute-list">
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">fitcons2_e067</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: FitCons2_E067</div>
            <p>The scores is the probability for a mutation to have a fitenss consequence.</p>

<p><img src="https://grr.seqpipe.org/hg19/scores/FitCons2_E067/statistics/histogram_FitCons2_E067.png" alt="HISTOGRAM" /></p>

<p>small values: more likely to have fitness consequence,
large_values less likely to have fitness consequence</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

          </div>
        </div>
        
      </div>

      <div class="annotator-line">
        <div class="annotator">
          <div class="annotator-title">Annotator type: position_score</div>
          
          <div class="resource-id">Resource id: <a href="https://grr.seqpipe.org/hg19/scores/FitCons2_E067/index.html">
                  hg19/scores/FitCons2_E067</a></div>                            
          
        </div>
  
        
        <div class="annotator-description">
          <ul>
<li><p>Annotator to use with genomic scores depending on genomic position like
phastCons, phyloP, FitCons2, etc.</p></li>
<li><p><a href="https://www.iossifovlab.com/gpfuserdocs/administration/annotation_tools.html#position-score" target="_blank">More info</a></p></li>
<li><p><strong>input_annotatable</strong>: <code>hg19_annotatable</code></p></li>
</ul>

        </div>
        

        
        <div class="resource">
          <div><b>Resource</b></div>
          <div>id: <a href="https://grr.seqpipe.org/hg19/scores/FitCons2_E067/index.html">
                  hg19/scores/FitCons2_E067</a></div>                            
          <div>type: position_score</div>
          
            <div>Description:</div>
            <div class="markdown">
              <p>FitCons2 score computed for the Brain Angular Gyrus (E067) tissue.</p>

            </div>
          
        </div>
        
      </div>
  
    </div>
    
    <div class="annotator-doc">
      <div class="attribute-list">
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">fitcons2_e068</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: FitCons2_E068</div>
            <p>The scores is the probability for a mutation to have a fitenss consequence.</p>

<p><img src="https://grr.seqpipe.org/hg19/scores/FitCons2_E068/statistics/histogram_FitCons2_E068.png" alt="HISTOGRAM" /></p>

<p>small values: more likely to have fitness consequence,
large_values less likely to have fitness consequence</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

          </div>
        </div>
        
      </div>

      <div class="annotator-line">
        <div class="annotator">
          <div class="annotator-title">Annotator type: position_score</div>
          
          <div class="resource-id">Resource id: <a href="https://grr.seqpipe.org/hg19/scores/FitCons2_E068/index.html">
                  hg19/scores/FitCons2_E068</a></div>                            
          
        </div>
  
        
        <div class="annotator-description">
          <ul>
<li><p>Annotator to use with genomic scores depending on genomic position like
phastCons, phyloP, FitCons2, etc.</p></li>
<li><p><a href="https://www.iossifovlab.com/gpfuserdocs/administration/annotation_tools.html#position-score" target="_blank">More info</a></p></li>
<li><p><strong>input_annotatable</strong>: <code>hg19_annotatable</code></p></li>
</ul>

        </div>
        

        
        <div class="resource">
          <div><b>Resource</b></div>
          <div>id: <a href="https://grr.seqpipe.org/hg19/scores/FitCons2_E068/index.html">
                  hg19/scores/FitCons2_E068</a></div>                            
          <div>type: position_score</div>
          
            <div>Description:</div>
            <div class="markdown">
              <p>FitCons2 score computed for the Brain Anterior Caudate (E068) tissue.</p>

            </div>
          
        </div>
        
      </div>
  
    </div>
    
    <div class="annotator-doc">
      <div class="attribute-list">
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">fitcons2_e069</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: FitCons2_E069</div>
            <p>The scores is the probability for a mutation to have a fitenss consequence.</p>

<p><img src="https://grr.seqpipe.org/hg19/scores/FitCons2_E069/statistics/histogram_FitCons2_E069.png" alt="HISTOGRAM" /></p>

<p>small values: more likely to have fitness consequence,
large_values less likely to have fitness consequence</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

          </div>
        </div>
        
      </div>

      <div class="annotator-line">
        <div class="annotator">
          <div class="annotator-title">Annotator type: position_score</div>
          
          <div class="resource-id">Resource id: <a href="https://grr.seqpipe.org/hg19/scores/FitCons2_E069/index.html">
                  hg19/scores/FitCons2_E069</a></div>                            
          
        </div>
  
        
        <div class="annotator-description">
          <ul>
<li><p>Annotator to use with genomic scores depending on genomic position like
phastCons, phyloP, FitCons2, etc.</p></li>
<li><p><a href="https://www.iossifovlab.com/gpfuserdocs/administration/annotation_tools.html#position-score" target="_blank">More info</a></p></li>
<li><p><strong>input_annotatable</strong>: <code>hg19_annotatable</code></p></li>
</ul>

        </div>
        

        
        <div class="resource">
          <div><b>Resource</b></div>
          <div>id: <a href="https://grr.seqpipe.org/hg19/scores/FitCons2_E069/index.html">
                  hg19/scores/FitCons2_E069</a></div>                            
          <div>type: position_score</div>
          
            <div>Description:</div>
            <div class="markdown">
              <p>FitCons2 score computed for the Brain Cingulate Gyrus (E069) tissue.</p>

            </div>
          
        </div>
        
      </div>
  
    </div>
    
    <div class="annotator-doc">
      <div class="attribute-list">
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">fitcons2_e070</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: FitCons2_E070</div>
            <p>The scores is the probability for a mutation to have a fitenss consequence.</p>

<p><img src="https://grr.seqpipe.org/hg19/scores/FitCons2_E070/statistics/histogram_FitCons2_E070.png" alt="HISTOGRAM" /></p>

<p>small values: more likely to have fitness consequence,
large_values less likely to have fitness consequence</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

          </div>
        </div>
        
      </div>

      <div class="annotator-line">
        <div class="annotator">
          <div class="annotator-title">Annotator type: position_score</div>
          
          <div class="resource-id">Resource id: <a href="https://grr.seqpipe.org/hg19/scores/FitCons2_E070/index.html">
                  hg19/scores/FitCons2_E070</a></div>                            
          
        </div>
  
        
        <div class="annotator-description">
          <ul>
<li><p>Annotator to use with genomic scores depending on genomic position like
phastCons, phyloP, FitCons2, etc.</p></li>
<li><p><a href="https://www.iossifovlab.com/gpfuserdocs/administration/annotation_tools.html#position-score" target="_blank">More info</a></p></li>
<li><p><strong>input_annotatable</strong>: <code>hg19_annotatable</code></p></li>
</ul>

        </div>
        

        
        <div class="resource">
          <div><b>Resource</b></div>
          <div>id: <a href="https://grr.seqpipe.org/hg19/scores/FitCons2_E070/index.html">
                  hg19/scores/FitCons2_E070</a></div>                            
          <div>type: position_score</div>
          
            <div>Description:</div>
            <div class="markdown">
              <p>FitCons2 score computed for the Brain Germinal Matrix (E070) tissue.</p>

            </div>
          
        </div>
        
      </div>
  
    </div>
    
    <div class="annotator-doc">
      <div class="attribute-list">
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">fitcons2_e071</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: FitCons2_E071</div>
            <p>The scores is the probability for a mutation to have a fitenss consequence.</p>

<p><img src="https://grr.seqpipe.org/hg19/scores/FitCons2_E071/statistics/histogram_FitCons2_E071.png" alt="HISTOGRAM" /></p>

<p>small values: more likely to have fitness consequence,
large_values less likely to have fitness consequence</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

          </div>
        </div>
        
      </div>

      <div class="annotator-line">
        <div class="annotator">
          <div class="annotator-title">Annotator type: position_score</div>
          
          <div class="resource-id">Resource id: <a href="https://grr.seqpipe.org/hg19/scores/FitCons2_E071/index.html">
                  hg19/scores/FitCons2_E071</a></div>                            
          
        </div>
  
        
        <div class="annotator-description">
          <ul>
<li><p>Annotator to use with genomic scores depending on genomic position like
phastCons, phyloP, FitCons2, etc.</p></li>
<li><p><a href="https://www.iossifovlab.com/gpfuserdocs/administration/annotation_tools.html#position-score" target="_blank">More info</a></p></li>
<li><p><strong>input_annotatable</strong>: <code>hg19_annotatable</code></p></li>
</ul>

        </div>
        

        
        <div class="resource">
          <div><b>Resource</b></div>
          <div>id: <a href="https://grr.seqpipe.org/hg19/scores/FitCons2_E071/index.html">
                  hg19/scores/FitCons2_E071</a></div>                            
          <div>type: position_score</div>
          
            <div>Description:</div>
            <div class="markdown">
              <p>FitCons2 score computed for the Brain Hippocampus Middle (E071) tissue.</p>

            </div>
          
        </div>
        
      </div>
  
    </div>
    
    <div class="annotator-doc">
      <div class="attribute-list">
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">fitcons2_e072</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: FitCons2_E072</div>
            <p>The scores is the probability for a mutation to have a fitenss consequence.</p>

<p><img src="https://grr.seqpipe.org/hg19/scores/FitCons2_E072/statistics/histogram_FitCons2_E072.png" alt="HISTOGRAM" /></p>

<p>small values: more likely to have fitness consequence,
large_values less likely to have fitness consequence</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

          </div>
        </div>
        
      </div>

      <div class="annotator-line">
        <div class="annotator">
          <div class="annotator-title">Annotator type: position_score</div>
          
          <div class="resource-id">Resource id: <a href="https://grr.seqpipe.org/hg19/scores/FitCons2_E072/index.html">
                  hg19/scores/FitCons2_E072</a></div>                            
          
        </div>
  
        
        <div class="annotator-description">
          <ul>
<li><p>Annotator to use with genomic scores depending on genomic position like
phastCons, phyloP, FitCons2, etc.</p></li>
<li><p><a href="https://www.iossifovlab.com/gpfuserdocs/administration/annotation_tools.html#position-score" target="_blank">More info</a></p></li>
<li><p><strong>input_annotatable</strong>: <code>hg19_annotatable</code></p></li>
</ul>

        </div>
        

        
        <div class="resource">
          <div><b>Resource</b></div>
          <div>id: <a href="https://grr.seqpipe.org/hg19/scores/FitCons2_E072/index.html">
                  hg19/scores/FitCons2_E072</a></div>                            
          <div>type: position_score</div>
          
            <div>Description:</div>
            <div class="markdown">
              <p>FitCons2 score computed for the Brain Inferior Temporal Lobe (E072) tissue.</p>

            </div>
          
        </div>
        
      </div>
  
    </div>
    
    <div class="annotator-doc">
      <div class="attribute-list">
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">fitcons2_e073</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: FitCons2_E073</div>
            <p>The scores is the probability for a mutation to have a fitenss consequence.</p>

<p><img src="https://grr.seqpipe.org/hg19/scores/FitCons2_E073/statistics/histogram_FitCons2_E073.png" alt="HISTOGRAM" /></p>

<p>small values: more likely to have fitness consequence,
large_values less likely to have fitness consequence</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

          </div>
        </div>
        
      </div>

      <div class="annotator-line">
        <div class="annotator">
          <div class="annotator-title">Annotator type: position_score</div>
          
          <div class="resource-id">Resource id: <a href="https://grr.seqpipe.org/hg19/scores/FitCons2_E073/index.html">
                  hg19/scores/FitCons2_E073</a></div>                            
          
        </div>
  
        
        <div class="annotator-description">
          <ul>
<li><p>Annotator to use with genomic scores depending on genomic position like
phastCons, phyloP, FitCons2, etc.</p></li>
<li><p><a href="https://www.iossifovlab.com/gpfuserdocs/administration/annotation_tools.html#position-score" target="_blank">More info</a></p></li>
<li><p><strong>input_annotatable</strong>: <code>hg19_annotatable</code></p></li>
</ul>

        </div>
        

        
        <div class="resource">
          <div><b>Resource</b></div>
          <div>id: <a href="https://grr.seqpipe.org/hg19/scores/FitCons2_E073/index.html">
                  hg19/scores/FitCons2_E073</a></div>                            
          <div>type: position_score</div>
          
            <div>Description:</div>
            <div class="markdown">
              <p>FitCons2 score computed for the Brain Dorsolateral Prefrontal Cortex (E073) tissue.</p>

            </div>
          
        </div>
        
      </div>
  
    </div>
    
    <div class="annotator-doc">
      <div class="attribute-list">
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">fitcons2_e074</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: FitCons2_E074</div>
            <p>The scores is the probability for a mutation to have a fitenss consequence.</p>

<p><img src="https://grr.seqpipe.org/hg19/scores/FitCons2_E074/statistics/histogram_FitCons2_E074.png" alt="HISTOGRAM" /></p>

<p>small values: more likely to have fitness consequence,
large_values less likely to have fitness consequence</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

          </div>
        </div>
        
      </div>

      <div class="annotator-line">
        <div class="annotator">
          <div class="annotator-title">Annotator type: position_score</div>
          
          <div class="resource-id">Resource id: <a href="https://grr.seqpipe.org/hg19/scores/FitCons2_E074/index.html">
                  hg19/scores/FitCons2_E074</a></div>                            
          
        </div>
  
        
        <div class="annotator-description">
          <ul>
<li><p>Annotator to use with genomic scores depending on genomic position like
phastCons, phyloP, FitCons2, etc.</p></li>
<li><p><a href="https://www.iossifovlab.com/gpfuserdocs/administration/annotation_tools.html#position-score" target="_blank">More info</a></p></li>
<li><p><strong>input_annotatable</strong>: <code>hg19_annotatable</code></p></li>
</ul>

        </div>
        

        
        <div class="resource">
          <div><b>Resource</b></div>
          <div>id: <a href="https://grr.seqpipe.org/hg19/scores/FitCons2_E074/index.html">
                  hg19/scores/FitCons2_E074</a></div>                            
          <div>type: position_score</div>
          
            <div>Description:</div>
            <div class="markdown">
              <p>FitCons2 score computed for the Brain Substantia Nigra (E074) tissue.</p>

            </div>
          
        </div>
        
      </div>
  
    </div>
    
    <div class="annotator-doc">
      <div class="attribute-list">
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">fitcons2_e081</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: FitCons2_E081</div>
            <p>The scores is the probability for a mutation to have a fitenss consequence.</p>

<p><img src="https://grr.seqpipe.org/hg19/scores/FitCons2_E081/statistics/histogram_FitCons2_E081.png" alt="HISTOGRAM" /></p>

<p>small values: more likely to have fitness consequence,
large_values less likely to have fitness consequence</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

          </div>
        </div>
        
      </div>

      <div class="annotator-line">
        <div class="annotator">
          <div class="annotator-title">Annotator type: position_score</div>
          
          <div class="resource-id">Resource id: <a href="https://grr.seqpipe.org/hg19/scores/FitCons2_E081/index.html">
                  hg19/scores/FitCons2_E081</a></div>                            
          
        </div>
  
        
        <div class="annotator-description">
          <ul>
<li><p>Annotator to use with genomic scores depending on genomic position like
phastCons, phyloP, FitCons2, etc.</p></li>
<li><p><a href="https://www.iossifovlab.com/gpfuserdocs/administration/annotation_tools.html#position-score" target="_blank">More info</a></p></li>
<li><p><strong>input_annotatable</strong>: <code>hg19_annotatable</code></p></li>
</ul>

        </div>
        

        
        <div class="resource">
          <div><b>Resource</b></div>
          <div>id: <a href="https://grr.seqpipe.org/hg19/scores/FitCons2_E081/index.html">
                  hg19/scores/FitCons2_E081</a></div>                            
          <div>type: position_score</div>
          
            <div>Description:</div>
            <div class="markdown">
              <p>FitCons2 score computed for the Fetal Brain Male (E081) tissue.</p>

            </div>
          
        </div>
        
      </div>
  
    </div>
    
    <div class="annotator-doc">
      <div class="attribute-list">
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">fitcons2_e082</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: FitCons2_E082</div>
            <p>The scores is the probability for a mutation to have a fitenss consequence.</p>

<p><img src="https://grr.seqpipe.org/hg19/scores/FitCons2_E082/statistics/histogram_FitCons2_E082.png" alt="HISTOGRAM" /></p>

<p>small values: more likely to have fitness consequence,
large_values less likely to have fitness consequence</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

          </div>
        </div>
        
      </div>

      <div class="annotator-line">
        <div class="annotator">
          <div class="annotator-title">Annotator type: position_score</div>
          
          <div class="resource-id">Resource id: <a href="https://grr.seqpipe.org/hg19/scores/FitCons2_E082/index.html">
                  hg19/scores/FitCons2_E082</a></div>                            
          
        </div>
  
        
        <div class="annotator-description">
          <ul>
<li><p>Annotator to use with genomic scores depending on genomic position like
phastCons, phyloP, FitCons2, etc.</p></li>
<li><p><a href="https://www.iossifovlab.com/gpfuserdocs/administration/annotation_tools.html#position-score" target="_blank">More info</a></p></li>
<li><p><strong>input_annotatable</strong>: <code>hg19_annotatable</code></p></li>
</ul>

        </div>
        

        
        <div class="resource">
          <div><b>Resource</b></div>
          <div>id: <a href="https://grr.seqpipe.org/hg19/scores/FitCons2_E082/index.html">
                  hg19/scores/FitCons2_E082</a></div>                            
          <div>type: position_score</div>
          
            <div>Description:</div>
            <div class="markdown">
              <p>FitCons2 score computed for the Fetal Brain Female (E082) tissue.</p>

            </div>
          
        </div>
        
      </div>
  
    </div>
    
    <div class="annotator-doc">
      <div class="attribute-list">
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">mpc</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: MPC</div>
            <p>Missense badness, PolyPhen-2, and Constraint.
A deleteriousness prediction score for missense variants"</p>

<p><img src="https://grr.seqpipe.org/hg19/scores/MPC/statistics/histogram_MPC.png" alt="HISTOGRAM" /></p>

<p>small values: less pathogenic,
large_values more pathogenic</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

<p><strong>nucleotide_aggregator</strong>: <code>max</code> [default]</p>

          </div>
        </div>
        
      </div>

      <div class="annotator-line">
        <div class="annotator">
          <div class="annotator-title">Annotator type: np_score</div>
          
          <div class="resource-id">Resource id: <a href="https://grr.seqpipe.org/hg19/scores/MPC/index.html">
                  hg19/scores/MPC</a></div>                            
          
        </div>
  
        
        <div class="annotator-description">
          <ul>
<li><p>Annotator to use with genomic scores depending on genomic position and
nucleotide change like CADD, MPC, etc.</p></li>
<li><p><a href="https://www.iossifovlab.com/gpfuserdocs/administration/annotation_tools.html#np-score" target="_blank">More info</a></p></li>
<li><p><strong>input_annotatable</strong>: <code>hg19_annotatable</code></p></li>
</ul>

        </div>
        

        
        <div class="resource">
          <div><b>Resource</b></div>
          <div>id: <a href="https://grr.seqpipe.org/hg19/scores/MPC/index.html">
                  hg19/scores/MPC</a></div>                            
          <div>type: np_score</div>
          
            <div>Description:</div>
            <div class="markdown">
              <p>A deleteriousness prediction score for missense variants based on regional
missense constraint. The range of MPC score is 0 to 5. The larger the score, the
more likely the variant is pathogenic. </p>

<p>Given increasing numbers of patients who are undergoing exome or genome
sequencing, it is critical to establish tools and methods to interpret the
impact of genetic variation. While the ability to predict deleteriousness
for any given variant is limited, missense variants remain a particularly
challenging class of variation to interpret, since they can have drastically
different effects depending on both the precise location and specific
amino acid substitution of the variant. In order to better evaluate
missense variation, we leveraged the exome sequencing data of 60,706
individuals from the Exome Aggregation Consortium (ExAC) dataset to
identify sub-genic regions that are depleted of missense variation.
We further used this depletion as part of a novel missense deleteriousness
metric named MPC. We applied MPC to de novo missense variants and identified
a category of de novo missense variants with the same impact on
neurodevelopmental disorders as truncating mutations in intolerant
genes, supporting the value of incorporating regional missense constraint
in variant interpretation.</p>

<p>Details see doi: http://dx.doi.org/10.1101/148353.</p>

            </div>
          
        </div>
        
      </div>
  
    </div>
    
    <div class="annotator-doc">
      <div class="attribute-list">
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">normalized_allele</div>
            <div class="attribute-type">Type: annotatable</div>
            
            <div class="attribute-internal">Internal</div>
            
          </div>
          <div class="attribute-description">
            <div>source: normalized_allele</div>
            <p>Normalized allele.</p>

          </div>
        </div>
        
      </div>

      <div class="annotator-line">
        <div class="annotator">
          <div class="annotator-title">Annotator type: normalize_allele_annotator</div>
          
          <div class="resource-id">Resource id: <a href="https://grr.seqpipe.org/hg38/genomes/GRCh38-hg38/index.html">
                  hg38/genomes/GRCh38-hg38</a></div>                            
          
        </div>
  
        
        <div class="no-description">
          No description
        </div>
        

        
        <div class="resource">
          <div><b>Resource</b></div>
          <div>id: <a href="https://grr.seqpipe.org/hg38/genomes/GRCh38-hg38/index.html">
                  hg38/genomes/GRCh38-hg38</a></div>                            
          <div>type: genome</div>
          
            <div>Description:</div>
            <div class="markdown">
              <h2>HG38 Reference Genome</h2>

<p>Default HG38 reference genome used by GPF</p>

            </div>
          
        </div>
        
      </div>
  
    </div>
    
    <div class="annotator-doc">
      <div class="attribute-list">
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">ssc_freq</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: allele_frequency</div>
            <p>SFARI SSC WGS CSHL allele frequency in %</p>

<p><img src="https://grr.seqpipe.org/hg38/variant_frequencies/SSC_WG38_CSHL_2380/statistics/histogram_allele_frequency.png" alt="HISTOGRAM" /></p>

<p>small values: None,
large_values None</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

<p><strong>allele_aggregator</strong>: <code>max</code> [default]</p>

          </div>
        </div>
        
      </div>

      <div class="annotator-line">
        <div class="annotator">
          <div class="annotator-title">Annotator type: allele_score</div>
          
          <div class="resource-id">Resource id: <a href="https://grr.seqpipe.org/hg38/variant_frequencies/SSC_WG38_CSHL_2380/index.html">
                  hg38/variant_frequencies/SSC_WG38_CSHL_2380</a></div>                            
          
        </div>
  
        
        <div class="annotator-description">
          <ul>
<li><p>Annotator to use with scores that depend on allele like
variant frequencies, etc.</p></li>
<li><p><a href="https://www.iossifovlab.com/gpfuserdocs/administration/annotation_tools.html#allele-score" target="_blank">More info</a></p></li>
</ul>

        </div>
        

        
        <div class="resource">
          <div><b>Resource</b></div>
          <div>id: <a href="https://grr.seqpipe.org/hg38/variant_frequencies/SSC_WG38_CSHL_2380/index.html">
                  hg38/variant_frequencies/SSC_WG38_CSHL_2380</a></div>                            
          <div>type: allele_score</div>
          
            <div>Description:</div>
            <div class="markdown">
              <p>TODO</p>

<p>exported from SFARI<em>SSC</em>WGS_CSHL using 
<code>gpf_validation_data/data_hg38/exports/SFARI_SSC_WGS_CSHL_frequency</code> scripts</p>

            </div>
          
        </div>
        
      </div>
  
    </div>
    
    <div class="annotator-doc">
      <div class="attribute-list">
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">exome_gnomad_v2_1_1_af_percent</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: AF_percent</div>
            <p>Alternative allele frequency in the whole gnomAD exome samples v2.1.1 as %</p>

<p><img src="https://grr.seqpipe.org/hg38/variant_frequencies/gnomAD_v2.1.1_liftover/exomes/statistics/histogram_AF_percent.png" alt="HISTOGRAM" /></p>

<p>small values: None,
large_values None</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

<p><strong>allele_aggregator</strong>: <code>max</code> [default]</p>

          </div>
        </div>
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">exome_gnomad_v2_1_1_ac</div>
            <div class="attribute-type">Type: int</div>
            
          </div>
          <div class="attribute-description">
            <div>source: AC</div>
            <p>Alternative allele count in the whole gnomAD exome samples v2.1.1</p>

<p><img src="https://grr.seqpipe.org/hg38/variant_frequencies/gnomAD_v2.1.1_liftover/exomes/statistics/histogram_AC.png" alt="HISTOGRAM" /></p>

<p>small values: None,
large_values None</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

<p><strong>allele_aggregator</strong>: <code>max</code> [default]</p>

          </div>
        </div>
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">exome_gnomad_v2_1_1_af</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: AF</div>
            <p>Alternative allele frequency in the whole gnomAD exome samples v2.1.1</p>

<p><img src="https://grr.seqpipe.org/hg38/variant_frequencies/gnomAD_v2.1.1_liftover/exomes/statistics/histogram_AF.png" alt="HISTOGRAM" /></p>

<p>small values: None,
large_values None</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

<p><strong>allele_aggregator</strong>: <code>max</code> [default]</p>

          </div>
        </div>
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">exome_gnomad_v2_1_1_an</div>
            <div class="attribute-type">Type: int</div>
            
          </div>
          <div class="attribute-description">
            <div>source: AN</div>
            <p>Total allele count in the whole gnomAD exome samples v2.1.1</p>

<p><img src="https://grr.seqpipe.org/hg38/variant_frequencies/gnomAD_v2.1.1_liftover/exomes/statistics/histogram_AN.png" alt="HISTOGRAM" /></p>

<p>small values: None,
large_values None</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

<p><strong>allele_aggregator</strong>: <code>max</code> [default]</p>

          </div>
        </div>
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">exome_gnomad_v2_1_1_controls_ac</div>
            <div class="attribute-type">Type: int</div>
            
          </div>
          <div class="attribute-description">
            <div>source: controls_AC</div>
            <p>Alternative allele count in the controls subset of whole gnomAD exome samples v2.1.1</p>

<p><img src="https://grr.seqpipe.org/hg38/variant_frequencies/gnomAD_v2.1.1_liftover/exomes/statistics/histogram_controls_AC.png" alt="HISTOGRAM" /></p>

<p>small values: None,
large_values None</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

<p><strong>allele_aggregator</strong>: <code>max</code> [default]</p>

          </div>
        </div>
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">exome_gnomad_v2_1_1_controls_an</div>
            <div class="attribute-type">Type: int</div>
            
          </div>
          <div class="attribute-description">
            <div>source: controls_AN</div>
            <p>gnomAD v2.1.1 liftover exomes count of genotyped individuals in control group</p>

<p><img src="https://grr.seqpipe.org/hg38/variant_frequencies/gnomAD_v2.1.1_liftover/exomes/statistics/histogram_controls_AN.png" alt="HISTOGRAM" /></p>

<p>small values: None,
large_values None</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

<p><strong>allele_aggregator</strong>: <code>max</code> [default]</p>

          </div>
        </div>
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">exome_gnomad_v2_1_1_non_neuro_ac</div>
            <div class="attribute-type">Type: int</div>
            
          </div>
          <div class="attribute-description">
            <div>source: non_neuro_AC</div>
            <p>Alternative allele count in the non-neuro subset of whole gnomAD exome samples v2.1.1</p>

<p><img src="https://grr.seqpipe.org/hg38/variant_frequencies/gnomAD_v2.1.1_liftover/exomes/statistics/histogram_non_neuro_AC.png" alt="HISTOGRAM" /></p>

<p>small values: None,
large_values None</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

<p><strong>allele_aggregator</strong>: <code>max</code> [default]</p>

          </div>
        </div>
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">exome_gnomad_v2_1_1_non_neuro_an</div>
            <div class="attribute-type">Type: int</div>
            
          </div>
          <div class="attribute-description">
            <div>source: non_neuro_AN</div>
            <p>Total allele count in the non-neuro subset of whole gnomAD exome samples v2.1.1</p>

<p><img src="https://grr.seqpipe.org/hg38/variant_frequencies/gnomAD_v2.1.1_liftover/exomes/statistics/histogram_non_neuro_AN.png" alt="HISTOGRAM" /></p>

<p>small values: None,
large_values None</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

<p><strong>allele_aggregator</strong>: <code>max</code> [default]</p>

          </div>
        </div>
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">exome_gnomad_v2_1_1_controls_af_percent</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: controls_AF_percent</div>
            <p>Alternative allele frequency in the controls subset of whole gnomAD exome samples v2.1.1 as %</p>

<p><img src="https://grr.seqpipe.org/hg38/variant_frequencies/gnomAD_v2.1.1_liftover/exomes/statistics/histogram_controls_AF_percent.png" alt="HISTOGRAM" /></p>

<p>small values: None,
large_values None</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

<p><strong>allele_aggregator</strong>: <code>max</code> [default]</p>

          </div>
        </div>
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">exome_gnomad_v2_1_1_non_neuro_af_percent</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: non_neuro_AF_percent</div>
            <p>Alternative allele frequency in the non-neuro subset of whole gnomAD exome samples v2.1.1 as %</p>

<p><img src="https://grr.seqpipe.org/hg38/variant_frequencies/gnomAD_v2.1.1_liftover/exomes/statistics/histogram_non_neuro_AF_percent.png" alt="HISTOGRAM" /></p>

<p>small values: None,
large_values None</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

<p><strong>allele_aggregator</strong>: <code>max</code> [default]</p>

          </div>
        </div>
        
      </div>

      <div class="annotator-line">
        <div class="annotator">
          <div class="annotator-title">Annotator type: allele_score</div>
          
          <div class="resource-id">Resource id: <a href="https://grr.seqpipe.org/hg38/variant_frequencies/gnomAD_v2.1.1_liftover/exomes/index.html">
                  hg38/variant_frequencies/gnomAD_v2.1.1_liftover/exomes</a></div>                            
          
        </div>
  
        
        <div class="annotator-description">
          <ul>
<li><p>Annotator to use with scores that depend on allele like
variant frequencies, etc.</p></li>
<li><p><a href="https://www.iossifovlab.com/gpfuserdocs/administration/annotation_tools.html#allele-score" target="_blank">More info</a></p></li>
<li><p><strong>input_annotatable</strong>: <code>normalized_allele</code></p></li>
</ul>

        </div>
        

        
        <div class="resource">
          <div><b>Resource</b></div>
          <div>id: <a href="https://grr.seqpipe.org/hg38/variant_frequencies/gnomAD_v2.1.1_liftover/exomes/index.html">
                  hg38/variant_frequencies/gnomAD_v2.1.1_liftover/exomes</a></div>                            
          <div>type: allele_score</div>
          
            <div>Description:</div>
            <div class="markdown">
              <p>Liftover of gnomAD exomes v2.1.1 to hg38 published by the Broad Institute.</p>

            </div>
          
        </div>
        
      </div>
  
    </div>
    
    <div class="annotator-doc">
      <div class="attribute-list">
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">genome_gnomad_v2_1_1_af_percent</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: AF_percent</div>
            <p>Alternative allele frequency in the whole gnomAD genome samples v2.1.1 as %</p>

<p><img src="https://grr.seqpipe.org/hg38/variant_frequencies/gnomAD_v2.1.1_liftover/genomes/statistics/histogram_AF_percent.png" alt="HISTOGRAM" /></p>

<p>small values: None,
large_values None</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

<p><strong>allele_aggregator</strong>: <code>max</code> [default]</p>

          </div>
        </div>
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">genome_gnomad_v2_1_1_ac</div>
            <div class="attribute-type">Type: int</div>
            
          </div>
          <div class="attribute-description">
            <div>source: AC</div>
            <p>Alternative allele count in the whole gnomAD genome samples v2.1.1</p>

<p><img src="https://grr.seqpipe.org/hg38/variant_frequencies/gnomAD_v2.1.1_liftover/genomes/statistics/histogram_AC.png" alt="HISTOGRAM" /></p>

<p>small values: None,
large_values None</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

<p><strong>allele_aggregator</strong>: <code>max</code> [default]</p>

          </div>
        </div>
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">genome_gnomad_v2_1_1_af</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: AF</div>
            <p>Alternative allele frequency in the whole gnomAD genome samples v2.1.1</p>

<p><img src="https://grr.seqpipe.org/hg38/variant_frequencies/gnomAD_v2.1.1_liftover/genomes/statistics/histogram_AF.png" alt="HISTOGRAM" /></p>

<p>small values: None,
large_values None</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

<p><strong>allele_aggregator</strong>: <code>max</code> [default]</p>

          </div>
        </div>
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">genome_gnomad_v2_1_1_an</div>
            <div class="attribute-type">Type: int</div>
            
          </div>
          <div class="attribute-description">
            <div>source: AN</div>
            <p>Total allele count in the whole gnomAD genome samples v2.1.1</p>

<p><img src="https://grr.seqpipe.org/hg38/variant_frequencies/gnomAD_v2.1.1_liftover/genomes/statistics/histogram_AN.png" alt="HISTOGRAM" /></p>

<p>small values: None,
large_values None</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

<p><strong>allele_aggregator</strong>: <code>max</code> [default]</p>

          </div>
        </div>
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">genome_gnomad_v2_1_1_controls_ac</div>
            <div class="attribute-type">Type: int</div>
            
          </div>
          <div class="attribute-description">
            <div>source: controls_AC</div>
            <p>Alternative allele count in the controls subset of whole gnomAD genome samples v2.1.1</p>

<p><img src="https://grr.seqpipe.org/hg38/variant_frequencies/gnomAD_v2.1.1_liftover/genomes/statistics/histogram_controls_AC.png" alt="HISTOGRAM" /></p>

<p>small values: None,
large_values None</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

<p><strong>allele_aggregator</strong>: <code>max</code> [default]</p>

          </div>
        </div>
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">genome_gnomad_v2_1_1_controls_an</div>
            <div class="attribute-type">Type: int</div>
            
          </div>
          <div class="attribute-description">
            <div>source: controls_AN</div>
            <p>gnomAD v2.1.1 liftover genome count of genotyped individuals in control group</p>

<p><img src="https://grr.seqpipe.org/hg38/variant_frequencies/gnomAD_v2.1.1_liftover/genomes/statistics/histogram_controls_AN.png" alt="HISTOGRAM" /></p>

<p>small values: None,
large_values None</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

<p><strong>allele_aggregator</strong>: <code>max</code> [default]</p>

          </div>
        </div>
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">genome_gnomad_v2_1_1_non_neuro_ac</div>
            <div class="attribute-type">Type: int</div>
            
          </div>
          <div class="attribute-description">
            <div>source: non_neuro_AC</div>
            <p>Alternative allele count in the non-neuro subset of whole gnomAD genome samples v2.1.1</p>

<p><img src="https://grr.seqpipe.org/hg38/variant_frequencies/gnomAD_v2.1.1_liftover/genomes/statistics/histogram_non_neuro_AC.png" alt="HISTOGRAM" /></p>

<p>small values: None,
large_values None</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

<p><strong>allele_aggregator</strong>: <code>max</code> [default]</p>

          </div>
        </div>
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">genome_gnomad_v2_1_1_non_neuro_an</div>
            <div class="attribute-type">Type: int</div>
            
          </div>
          <div class="attribute-description">
            <div>source: non_neuro_AN</div>
            <p>Total allele count in the non-neuro subset of whole gnomAD genome samples v2.1.1</p>

<p><img src="https://grr.seqpipe.org/hg38/variant_frequencies/gnomAD_v2.1.1_liftover/genomes/statistics/histogram_non_neuro_AN.png" alt="HISTOGRAM" /></p>

<p>small values: None,
large_values None</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

<p><strong>allele_aggregator</strong>: <code>max</code> [default]</p>

          </div>
        </div>
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">genome_gnomad_v2_1_1_controls_af_percent</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: controls_AF_percent</div>
            <p>Alternative allele frequency in the controls subset of whole gnomAD genome samples v2.1.1 as %</p>

<p><img src="https://grr.seqpipe.org/hg38/variant_frequencies/gnomAD_v2.1.1_liftover/genomes/statistics/histogram_controls_AF_percent.png" alt="HISTOGRAM" /></p>

<p>small values: None,
large_values None</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

<p><strong>allele_aggregator</strong>: <code>max</code> [default]</p>

          </div>
        </div>
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">genome_gnomad_v2_1_1_non_neuro_af_percent</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: non_neuro_AF_percent</div>
            <p>Alternative allele frequency in the non-neuro subset of whole gnomAD genome samples v2.1.1 as %</p>

<p><img src="https://grr.seqpipe.org/hg38/variant_frequencies/gnomAD_v2.1.1_liftover/genomes/statistics/histogram_non_neuro_AF_percent.png" alt="HISTOGRAM" /></p>

<p>small values: None,
large_values None</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

<p><strong>allele_aggregator</strong>: <code>max</code> [default]</p>

          </div>
        </div>
        
      </div>

      <div class="annotator-line">
        <div class="annotator">
          <div class="annotator-title">Annotator type: allele_score</div>
          
          <div class="resource-id">Resource id: <a href="https://grr.seqpipe.org/hg38/variant_frequencies/gnomAD_v2.1.1_liftover/genomes/index.html">
                  hg38/variant_frequencies/gnomAD_v2.1.1_liftover/genomes</a></div>                            
          
        </div>
  
        
        <div class="annotator-description">
          <ul>
<li><p>Annotator to use with scores that depend on allele like
variant frequencies, etc.</p></li>
<li><p><a href="https://www.iossifovlab.com/gpfuserdocs/administration/annotation_tools.html#allele-score" target="_blank">More info</a></p></li>
<li><p><strong>input_annotatable</strong>: <code>normalized_allele</code></p></li>
</ul>

        </div>
        

        
        <div class="resource">
          <div><b>Resource</b></div>
          <div>id: <a href="https://grr.seqpipe.org/hg38/variant_frequencies/gnomAD_v2.1.1_liftover/genomes/index.html">
                  hg38/variant_frequencies/gnomAD_v2.1.1_liftover/genomes</a></div>                            
          <div>type: allele_score</div>
          
            <div>Description:</div>
            <div class="markdown">
              <h2>gnomAD genomes v2.1.1 liftover</h2>

<p>Original gnomAD genomes v2.1.1 liftover is downloaded on October 19, 2020
from https://gnomad.broadinstitute.org/.</p>

<p>tabix -s 1 -b 2 -e 2 -f gnomad..r2.1.1.extract.tsv.gz</p>

            </div>
          
        </div>
        
      </div>
  
    </div>
    
    <div class="annotator-doc">
      <div class="attribute-list">
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">genome_gnomad_v3_af_percent</div>
            <div class="attribute-type">Type: float</div>
            
          </div>
          <div class="attribute-description">
            <div>source: AF_percent</div>
            <p>Alternative allele frequency represented as a percent in the all gnomAD v3.0 genome samples.</p>

<p><img src="https://grr.seqpipe.org/hg38/variant_frequencies/gnomAD_v3/genomes/statistics/histogram_AF_percent.png" alt="HISTOGRAM" /></p>

<p>small values: None,
large_values None</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

<p><strong>allele_aggregator</strong>: <code>max</code> [default]</p>

          </div>
        </div>
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">genome_gnomad_v3_ac</div>
            <div class="attribute-type">Type: int</div>
            
          </div>
          <div class="attribute-description">
            <div>source: AC</div>
            <p>Number of alternative alleles in the all gnomAD v3.0 genome samples.</p>

<p><img src="https://grr.seqpipe.org/hg38/variant_frequencies/gnomAD_v3/genomes/statistics/histogram_AC.png" alt="HISTOGRAM" /></p>

<p>small values: None,
large_values None</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

<p><strong>allele_aggregator</strong>: <code>max</code> [default]</p>

          </div>
        </div>
        
        <div class="attribute">
          <div class="attribute-info">
            <div class="attribute-name">genome_gnomad_v3_an</div>
            <div class="attribute-type">Type: int</div>
            
          </div>
          <div class="attribute-description">
            <div>source: AN</div>
            <p>Number of genotyped individuals in the all gnomAD v3.0 genome samples.</p>

<p><img src="https://grr.seqpipe.org/hg38/variant_frequencies/gnomAD_v3/genomes/statistics/histogram_AN.png" alt="HISTOGRAM" /></p>

<p>small values: None,
large_values None</p>

<p><strong>position_aggregator</strong>: <code>mean</code> [default]</p>

<p><strong>allele_aggregator</strong>: <code>max</code> [default]</p>

          </div>
        </div>
        
      </div>

      <div class="annotator-line">
        <div class="annotator">
          <div class="annotator-title">Annotator type: allele_score</div>
          
          <div class="resource-id">Resource id: <a href="https://grr.seqpipe.org/hg38/variant_frequencies/gnomAD_v3/genomes/index.html">
                  hg38/variant_frequencies/gnomAD_v3/genomes</a></div>                            
          
        </div>
  
        
        <div class="annotator-description">
          <ul>
<li><p>Annotator to use with scores that depend on allele like
variant frequencies, etc.</p></li>
<li><p><a href="https://www.iossifovlab.com/gpfuserdocs/administration/annotation_tools.html#allele-score" target="_blank">More info</a></p></li>
<li><p><strong>input_annotatable</strong>: <code>normalized_allele</code></p></li>
</ul>

        </div>
        

        
        <div class="resource">
          <div><b>Resource</b></div>
          <div>id: <a href="https://grr.seqpipe.org/hg38/variant_frequencies/gnomAD_v3/genomes/index.html">
                  hg38/variant_frequencies/gnomAD_v3/genomes</a></div>                            
          <div>type: allele_score</div>
          
            <div>Description:</div>
            <div class="markdown">
              <p>gnomAD v3.0 variants built from ~150,000 samples with whole genome sequence data. </p>

            </div>
          
        </div>
        
      </div>
  
    </div>
    
  </body>
<html>
